src/main/kotlin/com/mdrobnak/lalrpop/psi/util/LpPsiUtil.kt

```diff
@@ -1,5 +1,63 @@
 package com.mdrobnak.lalrpop.psi.util
 
-import com.mdrobnak.lalrpop.psi.LpGrammarParam
+import com.mdrobnak.lalrpop.psi.*
+import org.rust.lang.core.psi.ext.childrenWithLeaves
+import org.rust.lang.core.psi.ext.elementType
 
-fun LpGrammarParam.getName(): String = this.firstChild.text
\ No newline at end of file
+val LpGrammarParam.name: String
+    get() = this.firstChild.text
+
+val LpNonterminalName.nonterminal: LpNonterminal
+    get() = parent as LpNonterminal
+
+val LpSymbol.isExplicitlySelected: Boolean
+    get() = this.childrenWithLeaves.first().elementType == LpElementTypes.LESSTHAN
+
+
+val LpAlternative.selected: List<LpSymbol>
+    get() {
+        return this.children.filterIsInstance<LpSymbol>().selected
+    }
+
+val LpNonterminalRef.arguments: LpNonterminalArguments?
+    get() = this.nextSibling as? LpNonterminalArguments
+
+val LpNonterminalName.nonterminalParent: LpNonterminal
+    get() = this.parent as LpNonterminal
+
+val LpTypeRef.lifetime: String?
+    get() = this.childrenWithLeaves.find { it.elementType == LpElementTypes.LIFETIME }?.text
+
+val LpTypeRef.lifetimeOrInfer: String
+    get() = this.lifetime ?: "'_"
+
+val LpTypeRef.isRefMut: Boolean
+    get() = this.childrenWithLeaves.find { it.elementType == LpElementTypes.MUT } != null
+
+val List<LpSymbol>.selected: List<LpSymbol>
+    get() = if (this.any { it.isExplicitlySelected }) {
+        this.filter { it.isExplicitlySelected }
+    } else {
+        this
+    }
+
+fun List<LpSymbol>.computeType(arguments: List<NonterminalGenericArgument>): String {
+    val sel = selected
+    return sel.joinToString(
+        prefix = if (sel.size != 1) "(" else "",
+        postfix = if (sel.size != 1) ")" else ""
+    ) {
+        it.resolveType(arguments)
+    }
+}
+
+val LpAction.alternativeParent: LpAlternative
+    get() = this.parent as LpAlternative
+
+fun <T> LpRepeatOp.switch(question: T, plus: T, multiply: T): T =
+    when (this.childrenWithLeaves.first().elementType) {
+        LpElementTypes.QUESTION -> question
+        LpElementTypes.PLUS -> plus
+        LpElementTypes.MULTIPLY -> multiply
+        else -> TODO("Unreachable") //TODO: should be something like rust's `unreachable!()` but I have no idea how to tell the kotlin compiler about that
```

Comment:
The fact that only these three values will be possible is a runtime property, not a compile-time one (mostly). Unlike Rust's enums or Kotlin's sealed classes, it _could_ be possible that the first child has some other element type since `LpElementTypes` contains so many other possibilities. The restriction to these three element types is only maintained by the grammar and it's parser-generator. The best we can do is fail out with an error telling the programmer to fix the issue.